lib/fx/migration.rb

```diff
@@ -0,0 +1,31 @@
+require "rails"
+
+module Fx
+  # @api private
+  module Migration
+
+    def fx_migration_action(*arguments, &block)
```

Comment:
I'm not excited about the complexity this code brings, I wanted something cleaner where we leveraged the `ActiveRecord::Base.table_name_prefix` to coerce:

```
create_function :test_function
```

gjven that the table name prefix is `ActiveRecord::Base.table_name_prefix = "abc_"`, I'd expect that we named the function `abc_test_function`. How we do that, I'm less sure about. But I don't see this being the way we go about implementing this.

The same goes for the other `*_function` and `*_trigger` migration helpers.

This implementation feels too convoluted for something I'd like to maintain going forward.

Reply:
Hi @teoljungberg thanks for looking into it! I went with this approach because I felt that it's more similar to the way Activerecord deals with the migrations, and schema in which they apply the `table_name_prefix` (and/or `table_name_suffix`) when the migration is run and not on the generate or dump. 

But if you rather keep using the `method_missing` way, as you said we need to rename the generated **functions** and **triggers**. I'm not sure if it's going to feel so consistent with other parts of active record but we can explore it for sure. 

Make sense to work with this other approach on another branch and open another PR, or what do you have in mind?

Reply:
Maybe it is the case that this fits best into how activerecord handles migrations, as you have done here, maybe it doesn't. I don't know enough to make a judgement call yet. However, the code as submitted now is not stable enough for something I'd maintain, it's more procedural than OO.

If we back up a bit, what's the use-case of using the prefix-and-suffixes? What do we want to accomplish here? If I create a function named `testing1_testing2` and have the prefix set to `abc_` and suffix to `_def`, should the function be created as `abc_testing1_testing2_def` in the database?

From where I'd tackle this, I'd see how it fits into `create_function` (to take one of the two code paths - functions vs triggers)

https://github.com/teoljungberg/fx/blob/946cdccbd12333deb8f4566c9852b49c0231a618/lib/fx/statements/function.rb#L7-L46

And they all point to Definition, so maybe that's a good place to inline this?

But to go back up to the original point, I don't understand _why_ we should support it, what the use cases are, and how they are expected to behave. I'd like to see those things adressed in the PR description before we move forward.

Reply:
> Maybe it is the case that this fits best into how activerecord handles migrations, as you have done here, maybe it doesn't. I don't know enough to make a judgement call yet. However, the code as submitted now is not stable enough for something I'd maintain, it's more procedural than OO.

I understand and make sense to don't focus on how activerecord is dealing with it and focus on the approach that `fx` has.

> If we back up a bit, what's the use-case of using the prefix-and-suffixes? 

One use case to set the `config.active_record.table_name_prefix = ...` is for when different services share the same database, so if two services for instance has the table `users` they don't collide each other, because each has their own `prefix` (or `suffix`)

> What do we want to accomplish here?

That's a good question! so in the case where the `table_name_prefix` the generation of the files will not fail, but when you run de migration does. Because they could not found the  `prefix-...` **function** or **trigger**. So thinking about it, that's why I focused on the migration run and not the generator part.

> (3) If I create a function named testing1_testing2 and have the prefix set to abc_ and suffix to _def, should the function be created as abc_testing1_testing2_def in the database? 

Another good question that I'm not sure what makes more sense.

Overall I understand that this adds complexity and weirdness to the hole gem, and maybe it's not something that should be added to it.  So feel free to close this PR if you think the same, or if you rather add support to this issue, following your suggestion, make sense to define the strategy or how to proceed. As you said could be at the `Definition` part or somewhere like it, and also the (3) needs to be anwsered, right?

Open to whatever you think is best for the gem!